contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,274 @@
+
+            FlowDKG.nodeFinalSubmission = {}
+            for id in nodeIDs {
+                FlowDKG.nodeFinalSubmission[id] = []
+            }
+
+            FlowDKG.whiteboardMessages = []
+
+            FlowDKG.uniqueFinalSubmissions = []
+
+            FlowDKG.uniqueFinalSubmissionCount = {}
+
+            emit StartDKG()
+        }
+
+        pub fun endDKG() {
+            FlowDKG.dkgEnabled = false
+
+            emit EndDKG(finalSubmission: FlowDKG.dkgCompleted())
+        }
+    }
+
+    access(account) fun submissionsEqual(_ existingSubmission: [String], _ submission: [String]): Bool {
+
+        // If the submission length is different than the one being compared to, it is not equal
+        if submission.length != existingSubmission.length {
+            return false
+        }
+
+        var index = 0
+
+        // Check each key in the submiission to make sure that it matches
+        // the existing one
+        for key in submission {
+            // If a key length is incorrect, it is an invalid submission
+            if key.length != FlowDKG.submissionKeyLength {
+                panic("Submission key length is not correct!")
+            }
+
+            // if a key is different, stop checking this submission
+            // and move on to the next one
+            if key != existingSubmission[index] {
+                return false
+            }
+
+            index = index + 1
+        }
+
+        return true
+
+    }
+
+    pub fun participantIsRegistered(_ nodeID: String): Bool {
+        return FlowDKG.nodeFinalSubmission[nodeID] != nil
+    }
+
+    pub fun participantIsClaimed(_ nodeID: String): Bool? {
+        return FlowDKG.nodeClaimed[nodeID]
+    }
+
+    // specify last X messages in the transaction
+    pub fun getWhiteBoardMessages(): [Message] {
+        return self.whiteboardMessages
+    }
+
+    /// Returns whether this node has successfully submitted a final submission for this epoch.
+    /// Returns nil if the node is not a participant
+    /// Returns empty array if the node is a participant or hasn't submitted
+    /// Returns a submission array if the node has submitted
+    pub fun nodeHasSubmitted(_ nodeID: String): [String]? {
+        return self.nodeFinalSubmission[nodeID]
+    }
+
+    /// Get the list of all the consensus node IDs participating
+    pub fun getConsensusNodeIDs(): [String] {
+        return self.nodeFinalSubmission.keys
+    }
+
+    /// Get the array of all the unique final submissions
+    pub fun getFinalSubmissions(): [[String]] {
+        return self.uniqueFinalSubmissions
+    }
+
+    /// Returns the final set of keys if any one set of keys has more than 50% submissions
+    /// Returns nil if not found (incomplete)
+    pub fun dkgCompleted(): [String]? {
+
+        var index = 0
+
+        for submission in self.uniqueFinalSubmissions {
+            if self.uniqueFinalSubmissionCount[index]! > UInt64(self.nodeFinalSubmission.keys.length/2) {
```

Comment:
Looking back at the docs, this should be slightly different: `floor((n-1)/2)` rather than `floor(n/2)`

https://www.notion.so/dapperlabs/Smart-contract-based-DKG-29ad04d8996c4aae8019348ce49085a0#1be0b475e8994e398ff77baad91c8167

Reply:
`self.nodeFinalSubmission.keys.length` is actually equal to `n` here because it is a dictionary that has an entry reserved for every node. It might be good to change the name of it though, to avoid confusion. I'll try to think of something

Reply:
Ah okay, I got confused with the length of a node's submission. Thanks for clarifying.